### PR TITLE
optional load modes for iCE40

### PR DESCRIPTION
**Pull request by benediktibk**
This makes the two load modes for the iCE40 optional and represents therefore a workaround for #77983.

### CODE REVIEW

**Comment by benediktibk** [comment]
based upon #80846

File: drivers/fpga/fpga_ice40.c
Comment (josuah) [inline drivers/fpga/fpga_ice40.c:631]: How about using [2 different compatibles](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/maxim/ds18b20/ds18b20.c#L359-L369) both targeting this same driver?

That way, it might be possible to move all these build time asserts into devicetree bindings, while also enforcing that only one of them is chosen for every given instance.

**Review by josuah — DISMISSED** [review]
I really like the direction it is taking, but wonder if it would be interesting to split this in 2 drivers in the same file, using [two different compatibles](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/maxim/ds18b20/ds18b20.c#L359-L369):

```c
#define ICE40_FIELDS_COMMON(inst) \
		.creset = GPIO_DT_SPEC_INST_GET(inst, creset_gpios),               \
		.cdone = GPIO_DT_SPEC_INST_GET(inst, cdone_gpios),                 \
		.mhz_delay_count = FPGA_ICE40_MHZ_DELAY_COUNT(inst),               \
		.config_delay_us = FPGA_ICE40_CONFIG_DELAY_US(inst),               \
		.creset_delay_us = FPGA_ICE40_CRESET_DELAY_US(inst),               \
		.leading_clocks = FPGA_ICE40_LEADING_CLOCKS(inst),                 \
		.trailing_clocks = FPGA_ICE40_TRAILING_CLOCKS(inst),               \


#define ICE40_DEFINE_SPI(inst) \
	static struct fpga_ice40_data fpga_ice40_data_##inst;                      \
                                                                                   \
	static const struct fpga_ice40_config fpga_ice40_config_##inst = {         \
		ICE40_FIELDS_COMMON(inst)                                          \
		.bus = SPI_DT_SPEC_INST_GET(inst, SPI_WORD_SET(8) | SPI_TRANSFER_MSB, 0),\
		.load = fpga_ice40_load_spi};                                      \
                                                                                   \
	DEVICE_DT_INST_DEFINE(inst, fpga_ice40_init, NULL, &fpga_ice40_data_##inst,\
			      &fpga_ice40_config_##inst, POST_KERNEL, CONFIG_FPGA_INIT_PRIORITY,\
			      &fpga_ice40_api);

#define ICE40_DEFINE_GPIO(inst) \
	FPGA_ICE40_PINCTRL_DEFINE(inst);                                           \
                                                                                   \
	static struct fpga_ice40_data fpga_ice40_data_##inst;                      \
                                                                                   \
	static const struct fpga_ice40_config fpga_ice40_config_##inst = {         \
		ICE40_FIELDS_COMMON(inst)                                          \
		.clk = GPIO_DT_SPEC_INST_GET_OR(inst, clk_gpios, {0}),             \
		.pico = GPIO_DT_SPEC_INST_GET_OR(inst, pico_gpios, {0}),           \
		.set = FPGA_ICE40_GPIO_PINS(inst, gpios_set_reg),                  \
		.clear = FPGA_ICE40_GPIO_PINS(inst, gpios_clear_reg),              \
		.load = fpga_ice40_load_gpio,                                      \
		FPGA_ICE40_PINCTRL_CONFIG(inst)};                                  \
                                                                                   \
	DEVICE_DT_INST_DEFINE(inst, fpga_ice40_init, NULL, &fpga_ice40_data_##inst,\
			      &fpga_ice40_config_##inst, POST_KERNEL, CONFIG_FPGA_INIT_PRIORITY,\
			      &fpga_ice40_api);


#define DT_DRV_COMPAT lattice_ice40_spi
DT_INST_FOREACH_STATUS_OKAY(ICE40_DEFINE)
#undef DT_DRV_COMPAT

#define DT_DRV_COMPAT lattice_ice40_gpio
DT_INST_FOREACH_STATUS_OKAY(ICE40_DEFINE)
#undef DT_DRV_COMPAT
```

That is a breaking change, though, and the change you introduced are not: +1 with/without splitting the bindings.

**Comment by benediktibk** [comment]
> I really like the direction it is taking, but wonder if it would be interesting to split this in 2 drivers in the same file, using [two different compatibles](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/maxim/ds18b20/ds18b20.c#L359-L369):

Interesting idea, and the breaking change should not be a big issue as the driver is still experimental. @cfriedt Preferences from your side?

File: drivers/fpga/Kconfig.ice40
Comment (benediktibk) [reply drivers/fpga/Kconfig.ice40:29]: I think this is not a good idea, because theoretically it could be necessary to have both options available at the same time. And a Kconfig choice would make this exclusive.

**Comment by cfriedt** [comment]
Yeah, until we have better support for deinitializing SPI controllers, I guess anything is better than not having multi-instance.

Let's try two DT compats - why not?

**Comment by benediktibk** [comment]
> Yeah, until we have better support for deinitializing SPI controllers, I guess anything is better than not having multi-instance.
> 
> Let's try two DT compats - why not?

Okay, I will do so. But I'm gonna wait until #81212 is merged to avoid a conflict.

**Comment by benediktibk** [comment]
I've separated the drivers into a SPI and GPIO bitbang implementation, with two different compatibles. I've kept the original binding as it is to avoid necessary changes downstream.

**Comment by benediktibk** [comment]
Added a bullet point in the release notes for the change in the binding.

File: drivers/fpga/fpga_ice40_bitbang.c
Comment (josuah) [inline drivers/fpga/fpga_ice40_bitbang.c:269]: I do not remember seeing a way to specify min/max in devicetree, but `required: True` on the bindings guards against them not being undefined.
No harm in checking twice though.

File: dts/bindings/fpga/lattice,ice40-fpga-bitbang.yaml
Comment (josuah) [inline dts/bindings/fpga/lattice,ice40-fpga-bitbang.yaml:48]: It might be possible to put a `required: True` to each of the mandatory fields, and remove them from the `BUILD_ASSERT()`, as a way to catch the errors earlier. I do not see bugs in doing it this way though!

**Review by josuah — DISMISSED** [review]
Thank you very much for this effort. I could as well be doing it since I suggested the changes...

The situation seems very clear now. Alot of `#ifdef` and `BUILD_ASSERT()` removed.

**Review by cfriedt — DISMISSED** [review]

File: drivers/fpga/fpga_ice40_bitbang.c
Comment (benediktibk) [reply drivers/fpga/fpga_ice40_bitbang.c:269]: Those are actually not necessary anymore with a required, I will remove them.

Comment (benediktibk) [reply drivers/fpga/fpga_ice40_bitbang.c]: done

File: dts/bindings/fpga/lattice,ice40-fpga-bitbang.yaml
Comment (benediktibk) [reply dts/bindings/fpga/lattice,ice40-fpga-bitbang.yaml]: done

**Review by josuah — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**
